jvb/src/main/kotlin/org/jitsi/videobridge/cc/vp9/Vp9Picture.kt

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright @ 2019 8x8, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jitsi.videobridge.cc.vp9
+
+import org.jitsi.nlj.codec.vpx.VpxUtils.Companion.applyExtendedPictureIdDelta
+import org.jitsi.nlj.rtp.VideoRtpPacket
+import org.jitsi.nlj.rtp.codec.vp9.Vp9Packet
+import java.util.Collections
+import kotlin.collections.ArrayList
+
+/**
+ * Groups together some RTP/VP9 fields that refer to a specific incoming VP9
+ * picture, which may consist of multiple frames (of different spatial layers).
+ *
+ * Instances of this class are *NOT* thread safe.
+ *
+ * @author Jonathan Lennox
+ */
+class Vp9Picture(packet: Vp9Packet) {
+    val frames = ArrayList<Vp9Frame?>()
+
+    init {
+        val sid = if (packet.hasLayerIndices) packet.spatialLayerIndex else 0
+
+        setFrameAtSid(Vp9Frame(packet), sid)
+    }
+
+    fun frame(sid: Int) = frames.getOrNull(sid)
+
+    private fun setFrameAtSid(frame: Vp9Frame, sid: Int) {
+        if (sid >= frames.size) {
+            frames.addAll(Collections.nCopies(sid - frames.size, null))
+            frames.add(frame)
+        } else {
+            frames[sid] = frame
+        }
+    }
+
+    /**
+     * Return the first (lowest-sid, earliest in decoding order) frame.
+     * A valid picture must have at least one frame, so this will always return one.
+     */
+    private fun firstFrame(): Vp9Frame {
```

Comment:
This may not actually return a frame of the lowest possible sid for this picture, and could return different results over time, right?  So this is more "return the first frame we know about for this picture" ?

Reply:
Yes, I've clarified the comment.